### PR TITLE
AJ-955: just the backup changes from #280

### PR DESCRIPTION
**Pull request by davidangb**
This PR started by extracting the changes to the backup APIs from #280. It does NOT include anything from that other PR about WDS->WDS communication when in clone mode or integration with Leo. It's just the changes to the backup APIs.

Then, I responded to my own comments in #280 and tweaked the implementation to address those comments. This PR looks different than #280, most obviously:
* I changed the API endpoint signatures
* I changed the API responses to use a generic `Job` response; this sets us up for future work in making these APIs _and_ snapshot-import, PFB-import APIs asynchronous. This change cascaded a bunch of Java model changes.

Objective of this PR is to keep the backup changes and the cloning changes separately reviewable. Hopefully we can merge this one first and then that other PR becomes smaller!

### CODE REVIEW

File: service/src/main/java/org/databiosphere/workspacedataservice/dao/PostgresBackupDao.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    public Job<BackupResponse> getBackupStatus(UUID trackingId) {
+        try {
+            MapSqlParameterSource params = new MapSqlParameterSource("trackingId", trackingId);
+            return namedTemplate.query(
+                    "select id, status, error, createdtime, updatedtime, requester, filename, description from sys_wds.backup WHERE id = :trackingId", params, new BackupJobRowMapper()).get(0);
+        }
+        catch(Exception e) {
+            LOGGER.error("Unable to insert record into sys_wds.backup due to error {}.", e.getMessage());
+            return null;
+        }
```
Comment (davidangb) [inline service/src/main/java/org/databiosphere/workspacedataservice/dao/PostgresBackupDao.java:45]: better exception handling

File: service/src/main/java/org/databiosphere/workspacedataservice/service/model/BackupSchema.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
+package org.databiosphere.workspacedataservice.service.model;
+
+import org.databiosphere.workspacedataservice.shared.model.BackupRequest;
+import org.databiosphere.workspacedataservice.shared.model.job.JobStatus;
+
+import java.sql.Timestamp;
+import java.time.Instant;
+import java.util.UUID;
+
+public class BackupSchema {
```
Comment (davidangb) [inline service/src/main/java/org/databiosphere/workspacedataservice/service/model/BackupSchema.java:10]: can be removed?

File: service/src/main/java/org/databiosphere/workspacedataservice/shared/model/job/JobStatus.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package org.databiosphere.workspacedataservice.shared.model.job;
+
+public enum JobStatus {
+    INITIATED,
+    STARTED,
+    INPROGRESS,
+    COMPLETED,
+    ERROR,
+    CANCELLED
```
Comment (davidangb) [inline service/src/main/java/org/databiosphere/workspacedataservice/shared/model/job/JobStatus.java:9]: rename some of these?

File: service/src/main/java/org/databiosphere/workspacedataservice/workspacemanager/WorkspaceManagerDao.java [hunk 4]
```diff
@@ -30,18 +30,48 @@ public void createDataRepoSnapshotReference(SnapshotModel snapshotModel) {
...
+      UUID workspaceUUID = UUID.fromString(workspaceId);
+      ResourceList resourceList = resourceApi.enumerateResources(workspaceUUID, 0, 5, ResourceType.AZURE_STORAGE_CONTAINER, null);
+      // note: it is possible a workspace may have more than one storage container associated with it
+      // but currently there is no way to tell which one is the primary except for checking the actual container name
+      var storageUUID = extractResourceId(resourceList);
+      if(storageUUID != null) {
+        CreatedAzureStorageContainerSasToken sasBundle = azureResourceApi.createAzureStorageContainerSasToken(workspaceUUID, storageUUID, null, null, null, null);
+        return sasBundle.getUrl();
+      }
+      else throw new ApiException("Can't locate a storage resource matching workspace Id. ");
```
Comment (davidangb) [inline service/src/main/java/org/databiosphere/workspacedataservice/workspacemanager/WorkspaceManagerDao.java:64]: retries?

File: service/src/main/java/org/databiosphere/workspacedataservice/dao/PostgresBackupDao.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+            return null;
+        } else {
+            throw new RuntimeException("Unexpected error: %s rows found for backup status query".formatted(responses.size()));
+        }
+    }
+
+    @Override
+    public boolean backupExists(UUID trackingId) {
+        return Boolean.TRUE.equals(namedTemplate.queryForObject(
+                "select exists(select from sys_wds.backup WHERE id = :trackingId)",
```
[the remark this answers is not in the document]
Comment (davidangb) [reply service/src/main/java/org/databiosphere/workspacedataservice/dao/PostgresBackupDao.java:53]: because `namedTemplate.queryForObject(…` can return null, if you try to return it directly then IntelliJ yells at you:
Screenshot 2023-07-05 at 10 33 16 AM

I assume it will work fine in most cases, fwiw. I wonder if Sonar would complain about it too.

**Review by calypsomatic — APPROVED** [review]

File: service/src/main/java/org/databiosphere/workspacedataservice/dao/PostgresBackupDao.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (calypsomatic) [reply service/src/main/java/org/databiosphere/workspacedataservice/dao/PostgresBackupDao.java:53]: Gotcha.  Thanks!

File: service/src/main/java/org/databiosphere/workspacedataservice/controller/CloningController.java [hunk 6]
```diff
@@ -1,26 +1,42 @@
...
 
     @PostMapping("/backup/{version}")
-    public ResponseEntity<BackupResponse> createBackup(@PathVariable("version") String version) {
-        BackupResponse response = backupRestoreService.backupAzureWDS(storage, version);
-        var status = response.backupStatus() ? HttpStatus.OK : HttpStatus.INTERNAL_SERVER_ERROR;
-        return new ResponseEntity<>(response, status);
+    public ResponseEntity<Job<BackupResponse>> createBackup(@PathVariable("version") String version,
+                                                               @RequestBody BackupRequest backupRequest) {
+        UUID trackingId = UUID.randomUUID();
+        // TODO: make async
```
Comment (yuliadub) [inline service/src/main/java/org/databiosphere/workspacedataservice/controller/CloningController.java:31]: remove comment? :)

File: service/src/integrationTest/java/org/databiosphere/workspacedataservice/storage/LocalFileStorage.java [hunk 7]
```diff
@@ -19,19 +25,26 @@ public void streamOutputToBlobStorage(InputStream fromStream, String blobName) {
...
+                targetFile.delete();
+            } catch (Exception e) {
+                // could not clean up the temp file; don't fail the test due to this.
+            }
         }
     }
 
-    public void streamInputFromBlobStorage(OutputStream toStream, String blobName) {
+    public void streamInputFromBlobStorage(OutputStream toStream, String blobName, String workspaceId) {
         try(InputStream inStream = LocalFileStorage.class.getResourceAsStream("/backup-test.sql")) {
```
Comment (yuliadub) [inline service/src/integrationTest/java/org/databiosphere/workspacedataservice/storage/LocalFileStorage.java:39]: should we name this something similar to backup to keep naming consistent? something like "WDS-integrationTest-LocalFileStorage-" something?

File: service/src/main/java/org/databiosphere/workspacedataservice/dao/PostgresBackupDao.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    @WriteTransaction
+    public void updateBackupStatus(UUID trackingId, JobStatus status) {
+        namedTemplate.getJdbcTemplate().update("update sys_wds.backup SET status = ?, updatedtime = ? where id = ?",
+                status.name(), Timestamp.from(Instant.now()), trackingId);
+        LOGGER.info("Backup job {} is now {}", trackingId, status);
+    }
+
+    @Override
+    @WriteTransaction
+    public void saveBackupError(UUID trackingId, String error) {
```
Comment (yuliadub) [inline service/src/main/java/org/databiosphere/workspacedataservice/dao/PostgresBackupDao.java:76]: should we also set status to error with this update?

Comment (yuliadub) [reply service/src/main/java/org/databiosphere/workspacedataservice/dao/PostgresBackupDao.java:76]: or update status to Error after we call this, either way!

File: service/src/main/java/org/databiosphere/workspacedataservice/workspacemanager/WorkspaceManagerDao.java [hunk 9]
```diff
@@ -30,18 +34,51 @@ public void createDataRepoSnapshotReference(SnapshotModel snapshotModel) {
...
-              new DataRepoSnapshotAttributes()
-                  .instanceName(INSTANCE_NAME)
-                  .snapshot(snapshotModel.getId().toString()))
-              .metadata(
-                  new ReferenceResourceCommonFields()
-                      .cloningInstructions(CloningInstructionsEnum.REFERENCE)
-                      .name("%s_%s".formatted(snapshotModel.getName(), timeStamp))
-              ),
-          UUID.fromString(workspaceId));
+                      .snapshot(
```
Comment (yuliadub) [inline service/src/main/java/org/databiosphere/workspacedataservice/workspacemanager/WorkspaceManagerDao.java:37]: looks like this happened to you too, it added extra tabs here

Comment (yuliadub) [reply service/src/main/java/org/databiosphere/workspacedataservice/workspacemanager/WorkspaceManagerDao.java:37]: I swear I fixed those but they reverted back somehow

File: service/src/main/resources/static/swagger/openapi-docs.yaml [hunk 10]
```diff
@@ -882,16 +957,16 @@ components:
...
-          $ref: '#/components/schemas/diskSpaceComponentDetails'         
+          $ref: '#/components/schemas/diskSpaceComponentDetails'
     diskSpaceComponentDetails:
       type: object
       properties:
         total:
           type: string
         free:
-          type: float
+          type: number
```
Comment (yuliadub) [inline service/src/main/resources/static/swagger/openapi-docs.yaml:967]: were these intentional changes ? seem unrelated so wanted to double check

**Comment by yuliadub** [comment]
a few minor comments but otherwise look good! thanks for getting this in - I agree it will help make my PR easier to wrap up!

**Review by yuliadub — APPROVED** [review]

File: service/src/main/java/org/databiosphere/workspacedataservice/controller/CloningController.java (hunk 6 above)
Comment (davidangb) [reply service/src/main/java/org/databiosphere/workspacedataservice/controller/CloningController.java:31]: The comment is still relevant! I haven't done the async work yet, though this PR gets pretty far with it.

File: service/src/main/java/org/databiosphere/workspacedataservice/service/BackupRestoreService.java [hunk 11]
```diff
@@ -74,20 +98,30 @@ public BackupResponse backupAzureWDS(BackUpFileStorage storage, String version)
...
-            storage.streamOutputToBlobStorage(localProcessLauncher.getInputStream(), blobName);
+            backupDao.updateBackupStatus(trackingId, JobStatus.RUNNING);
+            LOGGER.info("Starting streaming backup to storage.");
+            storage.streamOutputToBlobStorage(localProcessLauncher.getInputStream(), blobName, String.valueOf(requesterWorkspaceId));
             String error = checkForError(localProcessLauncher);
+
             if (StringUtils.isNotBlank(error)) {
-                return new BackupResponse(false, error);
+                LOGGER.error("process error: {}", error);
+                backupDao.saveBackupError(trackingId, error);
```
[the remark this answers is not in the document]
Comment (davidangb) [reply service/src/main/java/org/databiosphere/workspacedataservice/service/BackupRestoreService.java:108]: that happens internally to `saveBackupError` (see my other reply) - perhaps this method needs a rename

File: service/src/integrationTest/java/org/databiosphere/workspacedataservice/storage/LocalFileStorage.java (hunk 7 above)
Comment (davidangb) [reply service/src/integrationTest/java/org/databiosphere/workspacedataservice/storage/LocalFileStorage.java]: changed!

File: service/src/main/java/org/databiosphere/workspacedataservice/workspacemanager/WorkspaceManagerDao.java (hunk 9 above)
Comment (davidangb) [reply service/src/main/java/org/databiosphere/workspacedataservice/workspacemanager/WorkspaceManagerDao.java]: reverted the whitespace changes.

File: service/src/main/resources/static/swagger/openapi-docs.yaml (hunk 10 above)
Comment (davidangb) [reply service/src/main/resources/static/swagger/openapi-docs.yaml:967]: yes, these are intentional - I should have added a comment. "float" is not legal openapi syntax, so I changed to "number".

File: service/src/main/java/org/databiosphere/workspacedataservice/dao/PostgresBackupDao.java (hunk 8 above)
Comment (davidangb) [reply service/src/main/java/org/databiosphere/workspacedataservice/dao/PostgresBackupDao.java]: renamed … naming is hard!

File: service/src/main/java/org/databiosphere/workspacedataservice/service/BackupRestoreService.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (davidangb) [reply service/src/main/java/org/databiosphere/workspacedataservice/service/BackupRestoreService.java]: renamed … naming is hard!

File: service/src/main/java/org/databiosphere/workspacedataservice/dao/PostgresBackupDao.java (hunk 8 above)
Comment (yuliadub) [reply service/src/main/java/org/databiosphere/workspacedataservice/dao/PostgresBackupDao.java]: ha wow how did I miss that (the line that updates status)! thank you for tweaking the naming!

Comment (yuliadub) [reply service/src/main/java/org/databiosphere/workspacedataservice/dao/PostgresBackupDao.java]: I guess now that I think about it, would it make sense to just to update the error message and status in one sql call vs 2?

Comment (davidangb) [reply service/src/main/java/org/databiosphere/workspacedataservice/dao/PostgresBackupDao.java]: we talked with faces and you weren't adamant about this. For anyone else reading my preference is to leave it as two calls so we're not copy/pasting code related to the status update, since the status update also sets last-updated time and writes a log entry. This isn't performance-sensitive code so I don't mind the two db queries.
